Add snapshot controller metrics

ggriffiths opened the pull request:
Signed-off-by: Grant Griffiths <grant@portworx.com>

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Adds snapshot controller metrics

**Which issue(s) this PR fixes**:
Fixes #142 

**Special notes for your reviewer**:
Pending work:
~~1. Fix metric tests in pkg/metrics~~
~~2. emit metrics for non-success statuses~~

**Does this PR introduce a user-facing change?**:
```release-note
The `--metrics-address` and `--metrics-path` flags have been added to support metrics for the snapshot-controller. To enable metrics, the user must configure a `--metrics-path` such as `:9102`. The default value for `--metrics-path` is the empty string, meaning this feature is disabled by default. The default value for `--metrics-path` is `/metrics`.
```

ggriffiths commented on cmd/snapshot-controller/main.go:
```diff
@@ -87,6 +92,25 @@ func main() {
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	if *metricsAddress != "" {
+		srv, err := metricsManager.StartMetricsEndpoint(*metricsPath, *metricsAddress, nil, wg)
```
I wasn't sure how the logger was intended to be passed down to the StartMetricsEndpoint? Don't we just use `klog`? Wasn't sure how to pass an instance of klog to the metrics pkg.

yuxiangqian commented:
/assign

yuxiangqian commented:
/cc. @msau42 @mattcary @xing-yang

xing-yang commented:
Do you need a rebase?  It seems that the unit test is hitting this bug: https://github.com/kubernetes-csi/external-snapshotter/pull/381?

ggriffiths commented:
pkg/metric tests seem to be flakey now. Looking into it now.

xing-yang commented:
/test pull-kubernetes-csi-external-snapshotter-1-19-on-kubernetes-1-19

xing-yang commented on cmd/snapshot-controller/main.go:
```diff
@@ -87,6 +92,25 @@ func main() {
+		defer func() {
+			err := srv.Close()
+			if err != nil {
+				klog.Errorf("fFailed to shutdown metrics server: %s", err.Error())
```
typo: fFailed -> Failed

xing-yang commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -416,6 +428,7 @@ func (ctrl *csiSnapshotCommonController) syncUnreadySnapshot(snapshot *crdv1.Vol
 		if err != nil {
 			// snapshot is bound but content is not bound to snapshot correctly
 			ctrl.updateSnapshotErrorStatusWithEvent(snapshot, v1.EventTypeWarning, "SnapshotBindFailed", fmt.Sprintf("Snapshot failed to bind VolumeSnapshotContent, %v", err))
+			ctrl.RecordSnapshotMetrics(snapshot, createSnapshotOperationName, preProvisionedSnapshotType, err)
```
What if it returns at line 424?  Do we need to record metrics with err there?

xing-yang commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -455,11 +479,14 @@ func (ctrl *csiSnapshotCommonController) syncUnreadySnapshot(snapshot *crdv1.Vol
 	if snapshot.Spec.Source.PersistentVolumeClaimName == nil {
 		ctrl.updateSnapshotErrorStatusWithEvent(snapshot, v1.EventTypeWarning, "SnapshotPVCSourceMissing", fmt.Sprintf("PVC source for snapshot %s is missing", uniqueSnapshotName))
-		return fmt.Errorf("expected PVC source for snapshot %s but got nil", uniqueSnapshotName)
+		err := fmt.Errorf("expected PVC source for snapshot %s but got nil", uniqueSnapshotName)
```
Should we record something when it return nil at line 476?

xing-yang commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -425,27 +438,38 @@ func (ctrl *csiSnapshotCommonController) syncUnreadySnapshot(snapshot *crdv1.Vol
 	}
+
 	// snapshot.Spec.Source.VolumeSnapshotContentName == nil - dynamically creating snapshot
+	ctrl.StartMetricsOperation(snapshot, snapshottingOperationName, dynamicSnapshotType)
```
Do we also need to record error for snapshottingOperationName?  It looks like this only records a completion?

ggriffiths replied:
I couldn't find any error cases for `snapshottingOperationName` that didn't overlap with createSnapshot. Will take another look.

yuxiangqian commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -425,27 +438,38 @@ func (ctrl *csiSnapshotCommonController) syncUnreadySnapshot(snapshot *crdv1.Vol
 	}
+
 	// snapshot.Spec.Source.VolumeSnapshotContentName == nil - dynamically creating snapshot
+	ctrl.StartMetricsOperation(snapshot, snapshottingOperationName, dynamicSnapshotType)
```
I am not sure whether the timer should be started here.
line 451 to 477 is really to handle the situations when the controller crashed/restarted after a content has been created however binding has not been successfully executed.
Or is it your desire to only recode error counts? then at line 475 it should clean the operation, otherwise it would be memory leak?

ggriffiths replied:
I am not sure either. I thought this code block inside of syncUnreadySnapshot was the extent of pre-provisioned snapshot handling:
https://github.com/kubernetes-csi/external-snapshotter/pull/409/files#diff-1a3473f9016f7abcf5007414ff7141e39032aaf7aba079eeef262d04167208a2R423-R433

Is there another area where we handle this case?

yuxiangqian commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -1099,6 +1127,19 @@ func (ctrl *csiSnapshotCommonController) updateSnapshotStatus(snapshot *crdv1.Vo
 		}
+
+		// If this is the first time we're setting creation time, we know snapshotting has finished
+		if snapshotObj.Status != nil && snapshotObj.Status.CreationTime == nil &&
```
snapshotObj.Status can be nil
also, can use newStatus directly(it should not be nil) instead of checking snapshotClone.Status != nil

yuxiangqian commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -1186,6 +1227,22 @@ func (ctrl *csiSnapshotCommonController) getSnapshotClass(className string) (*cr
 	return class, nil
 }
 
+// getSnapshotClass is a helper function to get snapshot class from the class name.
```
comment does not match

yuxiangqian commented on pkg/common-controller/snapshot_metrics.go:
```diff
@@ -0,0 +1,77 @@
+package common_controller
```
I am not convinced we need to have this wrapper. OperationStatus definitely needs an implementation, however, I am not sure about the "StartMetricsOperation" and "RecordSnapshotMetrics" calls.  Those are public methods and not called in anywhere else.

ggriffiths replied:
Removed these two public wrapper functions

yuxiangqian commented on pkg/common-controller/snapshot_metrics.go:
```diff
@@ -0,0 +1,77 @@
+	}
+}
+
+func (sos SnapshotOperationStatus) String() string {
```
this implementation will likely create a LOT of time series of latency metrics (one for each error message) which is problematic. see:
https://prometheus.io/docs/practices/naming/#labels

the status string is a label of the latency metrics, we definitely need to limit the cardinality 
https://github.com/kubernetes-csi/external-snapshotter/blob/master/pkg/metrics/metrics.go#L134

ggriffiths replied:
Added a type `snapshotStatusType` with four types as per the design doc:
1. Success
2. Invalid request
3. Controller error
4. Storage system error

However, I couldn't find any examples of storage-system-error? Wouldn't that mainly be in the snapshotter sidecar?

msau42 replied:
For storage system error, are we able to get csi error code?

ggriffiths replied:
We could check the VolumeSnapshot.Status.Error, but it seems like that error is used for both storage-system and controller errors. It formats the message as a string, I believe losing the csi error code.
https://github.com/kubernetes-csi/external-snapshotter/blob/23b415b6aaa7e0eba402bc3984ae2b726b101a80/pkg/sidecar-controller/snapshot_controller.go#L141-L156

So I'm not sure we can get the csi error code from that error, unless you are talking about checking in another spot.

yuxiangqian commented on pkg/metrics/metrics.go:
```diff
@@ -95,6 +100,9 @@ type operationMetricsManager struct {
 	opLatencyMetrics *k8smetrics.HistogramVec
+
+	// opCountMetrics is a Counter metrics
+	opCountMetrics *k8smetrics.CounterVec
```
do we really need this? I thought HistogramVec already provides count? given that the metrics are labeled using status, a count can be easily achieved by doing a label = "success"/"error"?

ggriffiths replied:
Good point, removed the counter and added status to the histogram.

yuxiangqian commented on pkg/metrics/metrics_test.go:
```diff
@@ -165,7 +166,7 @@ snapshot_controller_operation_total_seconds_bucket{driver_name="driver",operatio
 snapshot_controller_operation_total_seconds_bucket{driver_name="driver",operation_name="should-have-been-added",operation_status="Success",le="600"} 1
 snapshot_controller_operation_total_seconds_bucket{driver_name="driver",operation_name="should-have-been-added",operation_status="Success",le="+Inf"} 1
-snapshot_controller_operation_total_seconds_sum{driver_name="driver",operation_name="should-have-been-added",operation_status="Success"} 0.3
+snapshot_controller_operation_total_seconds_sum{driver_name="driver",operation_name="should-have-been-added",operation_status="Success"} 0.1
```
curious why this is adjusted to lower?

ggriffiths replied:
Typo, will fix this.

yuxiangqian reviewed (commented):
Good job @ggriffiths

ggriffiths commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -237,6 +237,9 @@ func (ctrl *csiSnapshotCommonController) syncSnapshot(snapshot *crdv1.VolumeSnap
 	klog.V(5).Infof("processSnapshotWithDeletionTimestamp VolumeSnapshot[%s]: %s", utils.SnapshotKey(snapshot), utils.GetSnapshotStatusForLogging(snapshot))
 
+	// New snapshot CR, start operation metric
+	ctrl.StartMetricsOperation(snapshot, deleteSnapshotOperationName, "")
```
Added type for deletion metrics

ggriffiths commented on cmd/snapshot-controller/main.go:
```diff
@@ -110,6 +135,7 @@ func main() {
 		go ctrl.Run(*threads, stopCh)
 
 		// ...until SIGINT
+		defer wg.Done()
```
Moved to the defer func

ggriffiths commented:
Addressed most feedback @xing-yang @yuxiangqian ,

Pending work I'll do today:
- Rework status field to classify different error types, instead of exact error messages.
- Check for places to emit snapshotting operation failures

msau42 commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -252,6 +269,12 @@ func (ctrl *csiSnapshotCommonController) processSnapshotWithDeletionTimestamp(sn
 	// content has been found from content cache store
 	content, err := ctrl.getContentFromStore(contentName)
 	if err != nil {
+		ctrl.metricsManager.RecordMetrics(metrics.Operation{
```
In general, is it possible to have the `RecordMetrics` call as a defer right after Start? That way we don't need to remember to call RecordMetrics if we add more exit points in the future.

ggriffiths replied:
I think we could have some sort defer call to perform RecordMetrics, but we would still need to update the snapshot type before the defer is executed. We could initialize a snapshotStatusType and change it before the defer is executed, but we'd still need to remember to update that.

I think a solution could be to have a custom error type that implements in the error interface and requires a snapshotStatusType assigned along with it. This would allow us to emit the correct snapshotStatusType based on the error returned per function. That might be a little involved for the 1.20 code freeze though.

msau42 replied:
Is it possible to start the timer after we have figured out if it's dynamic or preprovisioned? We should already have the api object with the source that tells us this.

I think it's fine to have a field for status that needs to be updated before returning. What about initializing it to error, and only setting it to success in the one case?

ggriffiths replied:
Refactored this. The repetitive code is mainly from all of the error metrics emitting, so I added defer calls that only emit metrics if `err != nil`. Success metrics are still manually emitted outside of the defer calls, this is because they are mostly deeper in the call stack.

I initialized snapshotStatusType to be `controller-error`, and manually changed to `invalid-request`/etc when needed.

msau42 commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -236,6 +237,22 @@ func (ctrl *csiSnapshotCommonController) syncSnapshot(snapshot *crdv1.VolumeSnap
+	if snapshot.Spec.Source.VolumeSnapshotContentName != nil {
+		snapshotProvisionType = preProvisionedSnapshotType
+	}
+	ctrl.metricsManager.OperationStart(metrics.Operation{
```
Can we save the `metrics.Operation` struct in a variable and reuse it, so that we don't accidentally call `RecordMetrics` with the wrong parameters?

ggriffiths replied:
Good idea, will make this change.

msau42 commented on pkg/common-controller/snapshot_metrics.go:
```diff
@@ -0,0 +1,37 @@
+
+const (
+	createSnapshotOperationName = "CreateSnapshot"
+	snapshottingOperationName   = "Snapshotting"
```
What is the difference between snapshotting and createsnapshot?

ggriffiths replied:
CreateSnapshot is full end to end from VolumeSnapshot CR noticed until it's marked ready.
Snapshotting is defined as when the VolumeSnapshot CR is noticed until a VolumeSnapshotContent is created.

From the design doc:

Snapshotting:
* Start_time: controller notices the first time that there is a new VolumeSnapshot CR to dynamically provision a snapshot
* End_time: controller notices that the CR has a status with CreationTime field set to be non-nil

CreateSnapshot
* Start_time: controller notices the first time that there is a new VolumeSnapshot CR(both dynamic and pre-provisioned cases)
* End_time: controller notices that the CR has a status with Ready field set to be true

@yuxiangqian is it worth only having CreateSnapshot and DeleteSnapshot operations? What sort of use cases are there for needing the snapshotting operation?

msau42 replied:
My main ask here is to make the names more obvious what they're measuring. What about something like: "CreateSnapshot" and "CreateSnapshotAndReady"?

ggriffiths replied:
CreateSnapshotAndReady sounds good, that is more clear to me. @yuxiangqian does this sound good? 

In this case, we would actually rename Snapshotting to CreateSnapshot, and rename CreateSnapshot to CreateSnapshotAndReady:

CreateSnapshot:

* Start_time: controller notices the first time that there is a new VolumeSnapshot CR to dynamically provision a snapshot
* End_time: controller notices that the CR has a status with CreationTime field set to be non-nil

CreateSnapshotAndReady
* Start_time: controller notices the first time that there is a new VolumeSnapshot CR(both dynamic and pre-provisioned cases)
* End_time: controller notices that the CR has a status with Ready field set to be true

xing-yang replied:
Sounds good to me.

ggriffiths replied:
Changed the metric types to reflect the above.

msau42 commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -252,6 +269,12 @@ func (ctrl *csiSnapshotCommonController) processSnapshotWithDeletionTimestamp(sn
+			Driver:       driverName,
+			ResourceID:   snapshot.UID,
+			SnapshotType: string(snapshotProvisionType),
+		}, NewSnapshotOperationStatus(snapshotCodeControllerError))
```
Recording a metric completion with an error status is mixing the semantic of what this metric is supposed to measure, which is one operation end to end. Instead, what this logic is doing is recording a metric per sync loop, so for a single volume snapshot request, it's could potentially record multiple operations in the histogram for it.

If this is what we want, then we should change the description of the metric to reflect that and look into end to end metrics separately.

For end to end metrics, I'm wondering if actually all we need to do is compare with the object creation/deletion timestamps.

ggriffiths replied:
The goal of this metric is the following:
>Snapshot operation end to end duration in number of seconds. For example, the duration from a user creates a VolumeSnapshot to request taking a snapshot of a particular PVC to the VolumeSnapshot has a Status.Ready == true.

I like the idea of comparing `VolumeSnapshot.CreationTimestamp` with `VolumeSnapshot.Status.CreationTime` for the e2e operation length. 

We could have two metrics, one counter w/ status, and one e2e with the operation time. This is what I originally implemented. I'll let @xing-yang @yuxiangqian chime in on if just having the histogram is enough. I saved the counter metric code and easily can re-add it if needed.

msau42 replied:
I like to think of it as two different things to measure, and each should have its own set of metrics.

1) Total end to end time of an request. This covers the time from an end user perspective. A controller could retry multiple times, but from the user this is only one request. Intermediate errors are not a concern here.
2) Time of a controller reconciliation loop. This covers just one loop and is more for internal tracking of how efficient this controller is, and can offer some data on intermittent/intermediate errors.

The metric for each scenario could be a histogram that covers status and latency together. There is latency and status of a single control loop, and there is latency and status of a total end to end request (although the status is just success/cancel)

ggriffiths replied:
That sounds good to me. I'll change the PR soon to reflect the two different histograms.

xing-yang commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -546,6 +622,7 @@ func (ctrl *csiSnapshotCommonController) getDynamicallyProvisionedContentFromSto
 	if content.Spec.Source.VolumeHandle == nil {
```
Should we record a snapshotStatusTypeInvalidRequest when it returns at line 619?

```
	if content == nil {
		// no matching content with the desired name has been found in cache
		return nil, nil
	}
```

xing-yang commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -588,22 +666,34 @@ func (ctrl *csiSnapshotCommonController) getContentFromStore(contentName string)
 	// If PVC is not being deleted and finalizer is not added yet, a finalizer should be added to PVC until snapshot is created
 	klog.V(5).Infof("createSnapshotContent: Check if PVC is not being deleted and add Finalizer for source of snapshot [%s] if needed", snapshot.Name)
-	err := ctrl.ensurePVCFinalizer(snapshot)
+	err = ctrl.ensurePVCFinalizer(snapshot)
```
Record a snapshotStatusTypeControllerError when err != nil?

xing-yang commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -654,6 +744,7 @@ func (ctrl *csiSnapshotCommonController) createSnapshotContent(snapshot *crdv1.V
 		strerr := fmt.Sprintf("Error creating volume snapshot content object for snapshot %s: %v.", utils.SnapshotKey(snapshot), err)
 		klog.Error(strerr)
 		ctrl.eventRecorder.Event(snapshot, v1.EventTypeWarning, "CreateSnapshotContentFailed", strerr)
+		ctrl.metricsManager.RecordMetrics(createOperation, NewSnapshotOperationStatus(snapshotStatusTypeControllerError))
```
At line 701, record snapshotStatusTypeInvalidRequest?
```
	snapshotRef, err := ref.GetReference(scheme.Scheme, snapshot)
	if err != nil {
		return nil, err
	}
```

xing-yang commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -663,6 +754,7 @@ func (ctrl *csiSnapshotCommonController) createSnapshotContent(snapshot *crdv1.V
 	// Update content in the cache store
 	_, err = ctrl.storeContentUpdate(updateContent)
 	if err != nil {
+		ctrl.metricsManager.RecordMetrics(createOperation, NewSnapshotOperationStatus(snapshotStatusTypeControllerError))
```
I wonder if we need to record an error here.  We don't even return err.

ggriffiths replied:
Added return err and metric emit is handled by defer

xing-yang commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -1093,12 +1185,46 @@ func (ctrl *csiSnapshotCommonController) updateSnapshotStatus(snapshot *crdv1.Vo
 		}
+
+		// Must meet the following criteria to emit a successful snapshotting status
+		// 1. Previous status was nil OR Previous status had a nil
```
Previous status had a nil CreationTime

ggriffiths commented:
Ready for re-review, addressed most concerns and had one question for @yuxiangqian about one of the comments.

xing-yang commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -427,9 +458,25 @@ func (ctrl *csiSnapshotCommonController) syncUnreadySnapshot(snapshot *crdv1.Vol
+		snapshotStatusType = metrics.SnapshotStatusTypeSuccess
 		return nil
 	}
-	// snapshot.Spec.Source.VolumeSnapshotContentName == nil - dynamically creating snapshot
```
Can you not to remove this comment?

ggriffiths replied:
Oops, thought I added this when I added the metric. Just re-added.

xing-yang commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -533,19 +584,36 @@ func (ctrl *csiSnapshotCommonController) getPreprovisionedContentFromStore(snaps
 	}
 	if content == nil {
 		// no matching content with the desired name has been found in cache
+		ctrl.metricsManager.RecordMetrics(createAndReadyOperation, metrics.NewSnapshotOperationStatus(metrics.SnapshotStatusTypeInvalidRequest), snapshot)
```
If you have `defer` above, you only need to set the snapshotStatusType?

ggriffiths replied:
The defer only handles cases where err != nil. Otherwise, the defer would create ControllerError metrics on every reconcile without an error (even though the operation might not be finished)

xing-yang replied:
Got it.

xing-yang commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -1093,12 +1179,36 @@ func (ctrl *csiSnapshotCommonController) updateSnapshotStatus(snapshot *crdv1.Vo
 			return nil, newControllerUpdateError(utils.SnapshotKey(snapshot), err.Error())
 		}
+
+		// Must meet the following criteria to emit a successful snapshotting status
```
snapshotting -> CreateSnapshot

xing-yang commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -1093,12 +1179,36 @@ func (ctrl *csiSnapshotCommonController) updateSnapshotStatus(snapshot *crdv1.Vo
+			ctrl.metricsManager.RecordMetrics(createOperation, metrics.NewSnapshotOperationStatus(metrics.SnapshotStatusTypeSuccess), newSnapshotObj)
+		}
+
+		// Must meet the following criteria to emit a successful CreateSnapshot status
```
CreateSnapshot -> CreateSnapshotAndReady

xing-yang commented on pkg/metrics/metrics.go:
```diff
@@ -23,20 +23,38 @@ import (
+
+	CreateSnapshotOperationName         = "CreateSnapshot"
+	CreateSnapshotAndReadyOperationName = "CreateSnapshotAndReady"
+	DeleteSnapshotOperationName         = "DeleteSnapshot"
```
Can you add comments to explain these 3 operations

xing-yang commented on pkg/metrics/metrics.go:
```diff
@@ -126,27 +168,71 @@ func (opMgr *operationMetricsManager) RecordMetrics(op Operation, status Operati
+	if strStatus == string(SnapshotStatusTypeSuccess) {
+		switch op.Name {
+		case CreateSnapshotAndReadyOperationName:
+			// time since user first created volumesnapshot
```
Add comments about "ready"

xing-yang commented:
I just have a few comments, otherwise it looks good overall.

@yuxiangqian @msau42 please take a look to see if you have more comments.

ggriffiths commented:
Addressed those @xing-yang. Also, I just remembered this requirement in the design doc:
> As a storage admin/provider, I want to the total size of all snapshots for capacity planning

I can get this value in as a label or new metric if we want (in this PR or a future one).

xing-yang commented:
> I just remembered this requirement in the design doc:
>> As a storage admin/provider, I want to the total size of all snapshots for capacity planning
> I can get this value in as a label or new metric if we want (in this PR or a future one).

I'm fine either way.  @yuxiangqian @msau42 what do you think?

xing-yang commented:
/lgtm

@yuxiangqian @msau42 can you take another look?

ggriffiths commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -236,6 +237,22 @@ func (ctrl *csiSnapshotCommonController) syncSnapshot(snapshot *crdv1.VolumeSnap
+	ctrl.metricsManager.OperationStart(deleteOperation)
+	snapshotStatusType := metrics.SnapshotStatusTypeControllerError
+	defer func() {
+		if err != nil {
```
That's true. I'll add a `metrics.SnapshotStatusTypeInProgress`. This will be the default status, and we can update to ControllerError/etc as needed.

This way, we'll always record a metric, but won't have ControllerError by default.

msau42 commented on pkg/metrics/metrics.go:
```diff
@@ -23,20 +23,54 @@ import (
+	// PreProvisionedSnapshotType represents a snapshot that is pre-provisioned
+	PreProvisionedSnapshotType = snapshotProvisionType("pre-provisioned")
+
+	SnapshotStatusTypeUnknown            = snapshotStatusType("unknown")
```
Can you add comments about which statuses are relevant to which metric?

msau42 commented on pkg/metrics/metrics.go:
```diff
@@ -126,27 +184,71 @@ func (opMgr *operationMetricsManager) RecordMetrics(op Operation, status Operati
+	// record total operation durations if operation is successful
+	if strStatus == string(SnapshotStatusTypeSuccess) {
+		switch op.Name {
+		case CreateSnapshotAndReadyOperationName:
```
does tracking e2e time for CreateSnapshot also seem useful?

ggriffiths replied:
good point, added this

msau42 commented on cmd/snapshot-controller/main.go:
```diff
@@ -48,6 +50,9 @@ var (
 	leaderElection          = flag.Bool("leader-election", false, "Enables leader election.")
 	leaderElectionNamespace = flag.String("leader-election-namespace", "", "The namespace where the leader election resource exists. Defaults to the pod namespace if not set.")
+
+	metricsAddress = flag.String("metrics-address", "", "The TCP network address where the prometheus metrics endpoint will listen (example: `:8080`). The default is empty string, which means metrics endpoint is disabled.")
```
We have a fix for the leader election library coming soon, and it's going to require a http endpoint for /healthz. Can we rename this field to be more generic like "http-endpoint"?

cc @verult who's working on the leader election fix

msau42 commented on pkg/metrics/metrics.go:
```diff
@@ -23,20 +23,62 @@ import (
+
+	// CreateSnapshotOperationName is the operation that tracks how long the controller takes to create a snapshot.
+	// Specifically, the operation metric is emitted based on the following timestamps:
+	// - Start_time: controller notices the first time that there is a new VolumeSnapshot CR to dynamically provision a snapshot
```
The start time description is relevant to the e2e metric, but not the reconcilation one. Can we clarify these comments

msau42 commented on pkg/metrics/metrics.go:
```diff
@@ -23,20 +23,62 @@ import (
+	// Individual reconciliations (reconciliation_total_seconds) also use this status.
+	SnapshotStatusTypeSuccess = snapshotStatusType("success")
+	// SnapshotStatusTypeCancel represents that a CreateSnapshot, CreateSnapshotAndReady,
+	// or DeleteSnapshot has been deleted before finishing.
```
are we actually able to detect a canceled delete?

ggriffiths replied:
The intention here was to send a cancel metric for the CreateSnapshotAndReady operation if a delete is issued prior to completion. 

In this case, we're sending a successful delete metric AND cancel CreateSnapshotAndReady metric.

msau42 commented on pkg/metrics/metrics.go:
```diff
@@ -126,27 +192,76 @@ func (opMgr *operationMetricsManager) RecordMetrics(op Operation, status Operati
+				// emit this metric as a custom cancellation
+				opMgr.opRequestLatencyMetrics.WithLabelValues(
+					op.Driver,
+					CreateSnapshotAndReadyOperationName,
```
Is there a cancel for CreateSnapshot?

ggriffiths replied:
Added this

msau42 replied:
not seeing it?

ggriffiths replied:
Ah sorry, thought I did. Added now above

msau42 reviewed (commented):
Thanks for working through my comments! I just have some minor ones remaining. 

@yuxiangqian please take another look

ggriffiths commented:
Thanks for the quick reviews! Addressed the recent comments.

yuxiangqian commented:
@ggriffiths I was off for couple of days, will get back on this by EOD.

yuxiangqian commented on pkg/metrics/metrics.go:
```diff
@@ -126,27 +192,91 @@ func (opMgr *operationMetricsManager) RecordMetrics(op Operation, status Operati
+	opMgr.opReconciliationLatencyMetrics.WithLabelValues(op.Driver, op.Name, op.SnapshotType, strStatus).Observe(reconciliationDuration)
+
+	// record total operation durations if operation is successful
+	if strStatus == string(SnapshotStatusTypeSuccess) {
```
This switch case seemed to be a bit over complicated to me as each case would do the same piece of work to record an end to end latency metric, propose to make the following changes:
`if success`
`    duration = time.Since(cached.StartTime)`
`    opMgr.opRequestLatencyMetrics.With....()
`

yuxiangqian commented on pkg/metrics/metrics.go:
```diff
@@ -126,27 +192,91 @@ func (opMgr *operationMetricsManager) RecordMetrics(op Operation, status Operati
+		switch op.Name {
+		case CreateSnapshotOperationName:
+			// time since user first created volumesnapshot until a volumesnapshot.Status.CreationTime is aded
+			operationDuration := time.Since(snapshot.CreationTimestamp.Time).Seconds()
```
it should be: `time.Since(cached-start-time)`. the creation timestamp of snapshot object does not mean the time when the controller sees it. In an extreme case, say when the snapshot controller has not been installed initiatively, the user creates a snapshot object, and x-hours later the snapshot controller got installed, we should *only* start counting when the controller gets to process the snapshot object.

msau42 replied:
I disagree here. I think the end to end metrics should reflect the user experience, which includes time spent in other components. If there's a bug in the snapshot controller such that it restarts frequently or is slow to process a snapshot request, we should capture that.

yuxiangqian replied:
I agree slowness should be captured, but the point is not related here, it will be captured actually.
Crash-looping is a different case though. If controller is crash looping, the user should have been notified by other manner not the e2e metrics.
The point is that the controller is not installed at all when a snapshot CR is created, in this case, using the creation time stamp of a snapshot CR as the start time is not reasonable to me. it will create a weird spiky point in the time-series, and I suspect this is the experience a user would want.
setting up a meeting to discuss further tomorrow.

xing-yang replied:
How is this kind of metric calculated for PVC creation?  Do we use creation timestamp of PVC as the the start time or do we use the time when the PV controller sees the PVC object?

msau42 replied:
I think there's value to have multiple metrics each capturing different levels of response, starting with the user experience and getting more fine grained to the controller, then the csi driver.

The higher level metrics can be used to define SLOs, whereas the finer grained metrics can be used to debug where something may be going wrong.

xing-yang replied:
Sure.  I think we should keep this metric and modify the description to reflect what it measures.
As a followup, add another metric that calculates the time between the controller observed start time and readyToUse.

yuxiangqian commented on pkg/metrics/metrics.go:
```diff
@@ -126,27 +192,91 @@ func (opMgr *operationMetricsManager) RecordMetrics(op Operation, status Operati
+
+		case CreateSnapshotAndReadyOperationName:
+			// time since user first created volumesnapshot until ReadyToUse is true
+			operationDuration := time.Since(snapshot.CreationTimestamp.Time).Seconds()
```
ditto, should use cached time

yuxiangqian commented on pkg/metrics/metrics.go:
```diff
@@ -126,27 +192,91 @@ func (opMgr *operationMetricsManager) RecordMetrics(op Operation, status Operati
+
+			// Still emit the delete operation, as we are both deleting the snapshot and cancelling it's creation.
+			// For this duration, we use the time since user marked volumesnapshot for deletion.
+			operationDuration := time.Since(snapshot.DeletionTimestamp.Time).Seconds()
```
ditto, should use cached timestamp

yuxiangqian commented on pkg/metrics/metrics.go:
```diff
@@ -23,20 +23,62 @@ import (
+	// PreProvisionedSnapshotType represents a snapshot that is pre-provisioned
+	PreProvisionedSnapshotType = snapshotProvisionType("pre-provisioned")
+
+	// The following metrics are used for the individual reconciliation metrics.
```
comment irrelevant?

yuxiangqian commented on pkg/metrics/metrics.go:
```diff
@@ -126,27 +192,91 @@ func (opMgr *operationMetricsManager) RecordMetrics(op Operation, status Operati
+
+	// record invidual reconciliation time
+	reconciliationDuration := time.Since(ts).Seconds()
+	opMgr.opReconciliationLatencyMetrics.WithLabelValues(op.Driver, op.Name, op.SnapshotType, strStatus).Observe(reconciliationDuration)
```
I am not sure this is the best place to implement reconciliation metrics.
the whole purpose of the cache is to reserve a starting point when there are multiple reconciliations. for a single reconciliation, we do not need any cached timestamp. If reconciliation latency is useful, then we should probably just do the duration calculation within controller logic. WDYT?

yuxiangqian commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -425,27 +460,47 @@ func (ctrl *csiSnapshotCommonController) syncUnreadySnapshot(snapshot *crdv1.Vol
 			return err
 		}
+
+		snapshotStatusType = metrics.SnapshotStatusTypeSuccess
```
at this point, the VolumeSnapshot may or may not have "ready" set to true yet. This simply means the controller has successfully bound the volume snapshot to volume snapshot content. Sidecar will call CSI "ListSnapshots" to update  content status, and then common controller will resync in next reconciliation. setting status to success here effectively ends the e2e latency reporting which is wrong IMO.

ggriffiths commented on pkg/metrics/metrics.go:
```diff
@@ -175,3 +305,25 @@ func (opMgr *operationMetricsManager) StartMetricsEndpoint(pattern, addr string,
+
+// SnapshotOperationStatus represents the status for a snapshot controller operation
+type SnapshotOperationStatus struct {
+	statusCode snapshotStatusType
```
I think the string simplifies things when it comes to recording the metrics. Otherwise, we would need to have a switch statement to get the string version of an enum. I can switch to that model if needed though.

ggriffiths commented:
Ready for re-review. I think there's some bug which causes failed snapshot.UpdateStatus API calls to still update the local snapshot cache. As a result, metrics are never emitted after a failed snapshot.UpdateStatus API call. This is because we look at the cached Snapshot to see if a metric needs to be recorded.

As a workaround, I'm now recording metrics just before `snapshot.UpdateStatus`:
https://github.com/kubernetes-csi/external-snapshotter/pull/409/files#diff-1a3473f9016f7abcf5007414ff7141e39032aaf7aba079eeef262d04167208a2R1128-R1148

yuxiangqian commented on pkg/metrics/metrics.go:
```diff
@@ -69,18 +109,39 @@ type MetricsManager interface {
 	Name string
-	// the name of the driver which executes the operation
+	// Driver is the driver name which executes the operation
 	Driver string
```
discussed offline.
Putting driver name into the key (operation) seems to bring quite some problems with little benefits.
The name of the operation + resource UID should already be unique enough.
same thing applies to snapshot type.
There is no guarantee that a valid VolumeSnapshotClass exists when a VolumeSnapshot CR is created for dynamic provisioning.

given that, I am proposing the following:

key ->
type Operation struct {
  operation name
  resource uid
}

value:
type val struct {
  start ts
  driver
  snapshot type
}

xing-yang commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -664,6 +693,7 @@ func (ctrl *csiSnapshotCommonController) createSnapshotContent(snapshot *crdv1.V
 	_, err = ctrl.storeContentUpdate(updateContent)
 	if err != nil {
 		klog.Errorf("failed to update content store %v", err)
+		return nil, err
```
+1

yuxiangqian commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -1186,6 +1237,49 @@ func (ctrl *csiSnapshotCommonController) getSnapshotClass(className string) (*cr
+
+	var driverName string
+	// Try to get from snapshot class first
+	if vs.Spec.VolumeSnapshotClassName != nil {
```
I think we might need to swap the logic like following:

get driver from content.Spec.Driver
if content not found
get driver from snapshot class

There are two major reasons:
1. content.Spec.Driver is populated by controller in dynamic provisioning case. while snapshot class can be deleted and recreated, the new driver will not be propagated to the content, however in VolumeSnapshot it holds only a reference
2. in preprovisioned case, driver is a required field for content while there could be NO snapshot class.
WDYT?

ggriffiths replied:
Agreed, moved this around to align with the above suggestion.

yuxiangqian reviewed (commented):
thanks @ggriffiths for the work!

yuxiangqian commented on pkg/metrics/metrics.go:
```diff
@@ -105,46 +171,95 @@ func NewMetricsManager() MetricsManager {
-func (opMgr *operationMetricsManager) RecordMetrics(op Operation, status OperationStatus) {
-	obj, exists := opMgr.cache.Load(op)
+// RecordMetrics emits operation metrics
+func (opMgr *operationMetricsManager) RecordMetrics(opKey OperationKey, opStatus OperationStatus, snapshot *crdv1.VolumeSnapshot, driverName string) {
```
rm "snapshot"? its no longer needed?

yuxiangqian commented on pkg/metrics/metrics.go:
```diff
@@ -105,46 +171,95 @@ func NewMetricsManager() MetricsManager {
+}
+
+// recordCancelMetric records a metric for a create operation that hasn't finished
+func (opMgr *operationMetricsManager) recordCancelMetric(obj interface{}, key OperationKey, duration float64) {
```
I think we should simply have "opMgr.opLatencyMetrics.WithLabelValues(
		val.Driver,
		key.Name,
		val.SnapshotType,
		string(SnapshotStatusTypeCancel)," in the RecordMetrics call. Cache management will happen and after this call again?

ggriffiths replied:
if we're deleting this snapshot, won't that cause a leak?

When we're record a delete metric, we must
1. Load Delete operation key, record metric, and delete key from cache
2. Load Create operation key, record cancel metric, and delete key from cache
3. Load CreateAndReady operation key, record metric, and delete key from cache.

Since the operation name is part of the key, we have three unique keys to delete (if Create/CreateAndReady keys still exist)

yuxiangqian replied:
oh, yeah you are right. I read it wrongly :)

yuxiangqian reviewed (commented):
some nits.
@msau42 PTAL, for per reconciliation metric, we will do a followup.

msau42 commented on pkg/common-controller/snapshot_controller_base.go:
```diff
@@ -407,6 +413,13 @@ func (ctrl *csiSnapshotCommonController) updateContent(content *crdv1.VolumeSnap
+		klog.Errorf("failed to getSnapshotDriverName while recording metrics for snapshot %q: %s", utils.SnapshotKey(snapshot), err)
+	} else {
+		deleteOperationKey := metrics.NewOperationKey(metrics.DeleteSnapshotOperationName, snapshot.UID)
+		ctrl.metricsManager.RecordMetrics(deleteOperationKey, metrics.NewSnapshotOperationStatus(metrics.SnapshotStatusTypeSuccess), driverName)
```
Similar question for the createsnapshot operations

ggriffiths replied:
Yes, I'm not sure how we can address that short of occasionally checking and cleaning up keys for snapshots that no longer exist. We could run that as a scheduled goroutine. 

How do we handle this in the snapshotStore and contentStore caches @yuxiangqian? Are those cache records cleared on force deletes?

yuxiangqian replied:
My understanding is even a volumesnapshot has been deleted by user forcefully, the controller would still receive an event from API server. The controller should still have the snapshot locally cached in 'snapshotStore' (controller crash will free any RAM allocated anyways), thus in this case there should be no leaking as line 420 will be executed?

@ggriffiths as I suggested previously to remove the "else" block for metric reporting though. otherwise if the snapshotclass has been deleted, it will be a leak.

msau42 replied:
Do you think it would be worth adding some metrics into the operation cache so we could potentially detect leaks? I'm worried this way of metrics caching is prone to leaking especially if we introduce/change logic flow in the future.

ggriffiths replied:
Yes, I think that could be a good idea. I'm thinking that might be a good followup issue, to keep this PR simple.

ggriffiths replied:
Let me know if that is okay @msau42 and I'll open a separate issue/PR to do that work. If not, I can add it sometime this week in this PR.

msau42 replied:
Followup is fine, thanks!

ggriffiths replied:
Sounds good, opened a new issue here: https://github.com/kubernetes-csi/external-snapshotter/issues/434

verult commented on cmd/snapshot-controller/main.go:
```diff
@@ -48,6 +50,9 @@ var (
 	leaderElection          = flag.Bool("leader-election", false, "Enables leader election.")
 	leaderElectionNamespace = flag.String("leader-election-namespace", "", "The namespace where the leader election resource exists. Defaults to the pod namespace if not set.")
+
+	httpEndpoint = flag.String("http-endpoint", "", "The TCP network address where the http server will listen (example: `:8080`). The default is empty string, which means the http endpoint is disabled.")
```
nit: could be more descriptive on what the server is, i.e:

"The TCP network address where the HTTP server for diagnostics, including metrics, will listen (example: `:8080`). The default is empty string, which means the server is disabled."

ggriffiths replied:
Thanks, added this in

ggriffiths commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -1095,10 +1136,30 @@ func (ctrl *csiSnapshotCommonController) updateSnapshotStatus(snapshot *crdv1.Vo
+		if !utils.IsSnapshotCreated(snapshotObj) && utils.IsSnapshotCreated(snapshotClone) {
+			createOperationKey := metrics.NewOperationKey(metrics.CreateSnapshotOperationName, snapshot.UID)
+			ctrl.metricsManager.RecordMetrics(createOperationKey, metrics.NewSnapshotOperationStatus(metrics.SnapshotStatusTypeSuccess), driverName)
+		}
```
Ah I see - yes this would cause a leak. For example, if the CreateSnapshot operation took 1 reconcile to record, and CreateSnapshotAndReady took 5 reconciles to record, we would have 4 leaked CreateSnapshot operations. We can drop the metric in this case if we determine the old and new snapshots are already created.

ggriffiths replied:
Added a case to handle this.

yuxiangqian replied:
should we do this in "syncUnreadySnapshot" instead of here? i.e., if a snapshot has been cut, do not add the "createOperationKey" operation instead of an explicit dropping. I think that is a more clean solution. WDYT? @ggriffiths

ggriffiths replied:
Moved this to start in syncUnreadySnapshot only when the snapshot has not been created yet.

ggriffiths commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -1095,10 +1136,30 @@ func (ctrl *csiSnapshotCommonController) updateSnapshotStatus(snapshot *crdv1.Vo
 		snapshotClone := snapshotObj.DeepCopy()
 		snapshotClone.Status = newStatus
+
+		// We need to record metrics before updating the status due to a bug causing cache entries after a failed UpdateStatus call.
```
I don't think so. Since this is the first time the controller has seen that the snapshot has been created or is ready, it is okay to record IMO.

The update status will go through on the next reconcile so that the end user sees the snapshot as ready. 

The problem here was that UpdateStatus failed, but the snapshotStore cache recorded readyToUse as true. On the next reconcile loop, this snapshot would hit syncReadySnapshot, and update status with readyToUse=true (without starting a new operation, because readyToUse=true).

ggriffiths commented:
/test pull-kubernetes-csi-external-snapshotter-1-18-on-kubernetes-1-18

yuxiangqian commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -1094,10 +1138,32 @@ func (ctrl *csiSnapshotCommonController) updateSnapshotStatus(snapshot *crdv1.Vo
+			ctrl.metricsManager.RecordMetrics(createOperationKey, metrics.NewSnapshotOperationStatus(metrics.SnapshotStatusTypeSuccess), driverName)
+		}
+
+		// If the old snapshot and new snapshot have both already been created, make sure to drop the CreateSnapshot operation.
```
this comment is no longer needed

ggriffiths replied:
Removed this

yuxiangqian reviewed (commented):
one nit and LGTM. thanks @ggriffiths

yuxiangqian commented:
/lgtm
/approve
